hw/dv/sv/tl_agent/tl_agent_cfg.sv

```
@@ -50,4 +104,61 @@ class tl_agent_cfg extends dv_base_agent_cfg;
   `uvm_object_utils_end
   `uvm_object_new
 
+  function void post_randomize();
+    // We want to randomize the initial value of last_a_source_released, while respecting any
+    // constraints on a_source in the tl_seq_item (and its extensions). So we do it here instead.
+    tl_seq_item dummy = tl_seq_item::type_id::create("dummy");
+  `DV_CHECK_MEMBER_RANDOMIZE_FATAL(a_source, dummy)
+    last_a_source_released = dummy.a_source;
+  endfunction
+
+  // Randomizes the a_source within the tl_seq_item with additional constriants.
+  //
+  // This function takes the tl_seq_item object to specifically randomize the a_source such that it
+  // does not match an existing one in the a_source_q. This randomization also takes into account
+  // any existing constriants on a_source that may exist in the class definition. The new a_source
+  // value is randomized to be equal to the last a_source released from the a_source_in_pending_reqs
+  // _q 20% of the time (weight is configurable).
+  virtual function void randomize_a_source(tl_seq_item item, bit invoked_from_driver = 1'b0);
+    bit use_last_a_source_released;
+
+    // Return if `invoked_from_driver` and `driver_sets_a_source` are opposite in value.
+    if (invoked_from_driver ^ driver_sets_a_source) return;
```

I might have misunderstood something here, but do we actually need this short-circuit? I'm a little concerned that code can set itself up wrong, then calling `randomize_a_source(my_item)` silently does nothing. If the call site can statically avoid doing this, should this actually be a fatal check?